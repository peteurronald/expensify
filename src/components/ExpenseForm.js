import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import styled from "styled-components";
import { ContentContainer, TextInput, TextArea, Form, FormError, Button, ActionBar} from '../styles/styled';

const RemoveButton = styled(Button)`
    background: #888;   
`;


const now = moment();
//console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };
    onRemove = () => {
        this.props.onRemove();
    };
    render() {
        return (
            <Form onSubmit={this.onSubmit}>
                <FormError>
                    {this.state.error && <p>{this.state.error}</p>}
                </FormError>
                <TextInput
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <TextInput
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <TextArea
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </TextArea>
                <ActionBar>
                    <Button>Save Expense</Button>
                    <RemoveButton onClick={this.onRemove}>Remove Expense</RemoveButton>
                 </ActionBar>
            </Form>
        )
    }
}