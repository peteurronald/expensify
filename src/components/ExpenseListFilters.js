import React from 'react';
import { connect } from 'react-redux';
import styled from "styled-components";
import {css} from "styled-components";
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'
import { ContentContainer, TextInput, Select} from '../styles/styled';


const InputGroup = styled.div`
  display: flex;
  flex-direction: row;      
  margin-bottom:  ${props => props.theme.ssize};
 
  @media (max-width: ${props => props.theme.desktopBreakpoint}) {
    display: flex;
    flex-direction: column; 
    margin-bottom:  ${props => props.theme.lsize};
  }  
`;


const Item = styled.div`
    margin-bottom: ${props => props.theme.ssize};
    @media (min-width: ${props => props.theme.desktopBreakpoint}) {
        margin-right: ${props => props.theme.ssize};
        margin-bottom: 0 ${props => props.theme.ssize} 0 0;
    }
`;


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDateChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };
    render() {
        return (
            <ContentContainer>
                <InputGroup>
                    <Item>
                        <TextInput type="text" placeholder="Search Expenses..." value={this.props.filters.text} onChange={this.onTextChange} />
                    </Item>
                    <Item>
                        <Select value={this.props.filters.sortBy} onChange={this.onSortChange} >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </Select>
                    </Item>
                    <Item>
                        <DateRangePicker
                            startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                            startDateId="startDateId" // PropTypes.string.isRequired,
                            endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                            endDateId="endDateId" // PropTypes.string.isRequired,
                            onDatesChange={this.onDateChange} // PropTypes.func.isRequired,
                            focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </Item>
                </InputGroup>
            </ContentContainer>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => ({
    filters: state.filters
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);