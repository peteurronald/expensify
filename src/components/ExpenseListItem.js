import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import styled from "styled-components";


const LinkItem = styled(Link) `
    border: 1px solid green;  
    border-top: none;
    display:flex;
    flex-direction: column;
    color: green;
    padding:  ${props => props.theme.ssize};
    text-decoration: none;
    transition: background .3s ease;
    &:hover{
        background:  ${props => props.theme.offWhite};
    }
    @media (min-width: ${props => props.theme.desktopBreakpoint}) {
        align-items:center;
        flex-direction: row; 
        justify-content: space-between;
        padding:  ${props => props.theme.msize};
    }  
`;

const ListTitle = styled.h3`
    margin: 0;
    word-break: break-all;
`;

const ListSubTitle = styled.span`
    color:  ${props => props.theme.grey};
    font-size: 1.4rem;
`;

const ListData = styled.h3`
    margin: ${props => props.theme.ssize} 0 0 0;
    @media (min-width: ${props => props.theme.desktopBreakpoint}) {
        margin: 0;
        padding-left: ${props => props.theme.ssize};
    }
`;


export default ({ id, description, amount, createdAt }) => (
    <LinkItem to={`/edit/${id}`}>
        <div>
            <ListTitle>{description}</ListTitle>
            <ListSubTitle>{moment(createdAt).format('MMMM  Do, YYYY')} </ListSubTitle>
        </div>
        <ListData>
            {numeral(amount / 100).format('$0,0.00')}
        </ListData>
    </LinkItem>
);
