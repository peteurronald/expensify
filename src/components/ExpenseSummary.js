import React from 'react';
import { connect } from 'react-redux';

import selectExpensesTotal from "../selectors/expenses-total";
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';
import styled from 'styled-components';
import { ContentContainer, LinkButton } from '../styles/styled';
import {prop} from 'styled-tools';

const PageHeader = styled.div`
    background: green;
    margin-bottom: ${props => props.theme.ssize};
    padding: ${props => props.theme.lsize} 0;
`;


const Title = styled.h3`
    font-weight: 100;
    margin: 0;
    color: white;  
    text-align: right;  
`;

const Span = styled.span`
    font-weight: bold;
`;


const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount > 1 ? 'expenses' : 'expense';
    const totalExpense = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <PageHeader>
            <ContentContainer>
                {expensesCount === 0 && <Title>No expenses</Title>}
                {expensesCount > 0 &&
                    <Title> Viewing <Span>{expensesCount}</Span> {expenseWord} totalling <Span>{totalExpense}</Span> </Title>}
            </ContentContainer>
            <ContentContainer>           
                <LinkButton to="/create">Add Expense</LinkButton>
            </ContentContainer>
        </PageHeader>
    );
};

export default ExpenseSummary;
