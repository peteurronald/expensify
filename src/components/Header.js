import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';
import styled from "styled-components";
import {ContentContainer, PageHeader, PageHeaderTitle} from '../styles/styled';


const HeaderContent= styled.div`
    align-items: center;
    display: flex;
    justify-content: space-between;
    padding: 1.2rem 0;
`;


const LogoutButton = styled.button`
    background: none;
    color: white;
    border: 2px solid ${props => props.theme.main};
`;


export const Header = ({startLogout}) => (
    <PageHeader>
        <ContentContainer>
           <HeaderContent>  
                 <Link  to="/dashboard"  >
                   <PageHeaderTitle>Expensify</PageHeaderTitle>   
                </Link>
                <LogoutButton onClick={startLogout}>Logout</LogoutButton>
            </HeaderContent>      
        </ContentContainer>
    </PageHeader>    
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});


export default connect(undefined, mapDispatchToProps)(Header);

