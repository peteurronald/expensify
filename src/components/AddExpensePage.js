import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { ContentContainer, SubTitle } from '../styles/styled';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <ContentContainer>
                    <SubTitle>Add Expense</SubTitle>
                </ContentContainer>
                <ContentContainer>
                    <ExpenseForm onSubmit={this.onSubmit} />
                </ContentContainer>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);