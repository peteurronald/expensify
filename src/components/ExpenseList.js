import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import styled from "styled-components";
import { ContentContainer} from '../styles/styled';


const ShowForMobile = styled.div`
  @media (min-width: ${props => props.theme.desktopBreakpoint}) {
    display:none;
  }  
`;
const ShowForDesktop = styled.div`
  @media (max-width: ${props => props.theme.desktopBreakpoint}) {
    display:none;
  }  
`;

const ListHeader = styled.div`
    background:  green;  
    color: white;
    display: flex;
    justify-content: space-between;
    padding: ${props => props.theme.ssize} ${props => props.theme.msize}
`;

const ListBody = styled.div`
    margin-bottom:  ${props => props.theme.msize};
    @media (min-width: ${props => props.theme.desktopBreakpoint}) {
        margin-bottom:  ${props => props.theme.lsize};
    }    
`;

const Message = styled.span`
    display: flex;
    align-items: center;
    justify-content: center;
    color: ${props => props.theme.grey};
    padding: ${props => props.theme.msize};
    &:hover{
        background:none;
    };
`;

export const ExpenseList = (props) => (
    <ContentContainer>
        <ListHeader>
            <ShowForMobile>Expenses</ShowForMobile>
            <ShowForDesktop>Expense</ShowForDesktop>
            <ShowForDesktop>Amount</ShowForDesktop>
        </ListHeader>
        <ListBody>
            {
                props.expenses.length === 0 ?
                    <div>
                        <Message>No expenses</Message>
                    </div>
                    : (
                        props.expenses.map((expense) => {
                            return <ExpenseListItem key={expense.id} {...expense} />;
                        })
                    )
            }
        </ListBody>
    </ContentContainer>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);
