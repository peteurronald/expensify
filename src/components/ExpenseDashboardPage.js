import React from 'react';
import {connect } from 'react-redux';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal  from "../selectors/expenses-total";
import ExpenseSummary from './ExpenseSummary';

export const ExpenseDashboardPage = (props) => {
    const count = props.expenses.length;
    const total =  selectExpensesTotal(props.expenses);
    return (
        <div>   
            <ExpenseSummary expensesCount={count} expensesTotal={total} />
            <ExpenseListFilters />
            <ExpenseList />
        </div>   
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }; 
};

export default connect(mapStateToProps)(ExpenseDashboardPage);
