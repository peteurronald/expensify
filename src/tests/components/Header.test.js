import React from 'react';
import {Header} from '../../components/Header'
import {shallow} from 'enzyme';


let startLogoutSpy, wrapper;

beforeEach(()=> {
    startLogoutSpy = jest.fn();
    wrapper = shallow(<Header startLogout={startLogoutSpy}/>);
});

test('should render Header correctly',()=> {
     expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click',()=> {
    wrapper.find('button').simulate('click');
    expect(startLogoutSpy).toHaveBeenCalled();
});


