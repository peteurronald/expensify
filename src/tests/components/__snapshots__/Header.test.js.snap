// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<styled.header>
  <styled.div>
    <styled.div>
      <Link
        replace={false}
        to="/dashboard"
      >
        <styled.h1>
          Expensify
        </styled.h1>
      </Link>
      <styled.button
        onClick={[Function]}
      >
        Logout
      </styled.button>
    </styled.div>
  </styled.div>
</styled.header>
`;
